dind/pkg/cluster/cluster.go

```diff
@@ -0,0 +1,359 @@
+	// Finds a non-localhost unicast address on the given interface.
+	intf, err := net.InterfaceByName(intfName)
+	if err != nil {
+		return "", err
+	}
+	addrs, err := intf.Addrs()
+	if err != nil {
+		return "", err
+	}
+	for _, addr := range addrs {
+		if !strings.HasPrefix("127", addr.String()) {
+			return addr.String(), nil
+		}
+	}
+
+	return "", fmt.Errorf("No non-localhost unicast addresses found on interface %s", intfName)
+}
+
+type Cluster struct {
+	numNodes         int
+	healthyNodes     int
+	apiserverHealthy bool
+
+	externalProxy   string
+	image           string
+	kubecfg         string
+	masterConfigDir string
+	networkID       string
+	networkName     string
+	nodeImageFile   string
+	version         string
+	nodeIDs         []string
+
+	apiserver *kubernetes.Clientset
+	docker    *client.Client
+	timeout   time.Duration
+}
+
+func New(numNodes int, externalProxy, image, masterConfigDir, networkName, nodeImageFile, version string, timeout time.Duration) (*Cluster, error) {
+	docker, err := client.NewEnvClient()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Cluster{
+		numNodes:        numNodes,
+		externalProxy:   externalProxy,
+		docker:          docker,
+		image:           image,
+		kubecfg:         masterConfigDir + "/admin.conf",
+		masterConfigDir: masterConfigDir,
+		networkName:     networkName,
+		nodeImageFile:   nodeImageFile,
+		timeout:         timeout,
+		version:         version,
+	}, nil
+}
+
+// Up synchronously starts a cluster, reporting progress.
+func (c *Cluster) Up(loadImage bool) error {
+	end := time.Now().Add(c.timeout)
+	if loadImage {
+		glog.V(2).Info("Loading node image")
+		if err := c.LoadImage(); err != nil {
+			return err
+		}
+	}
+	glog.V(2).Info("Creating docker objects")
+	if err := c.CreateDockerObjects(); err != nil {
+		return err
+	}
+	glog.V(2).Info("Starting containers")
+	if err := c.StartContainers(); err != nil {
+		return err
+	}
+	glog.V(2).Info("Waiting for apiserver")
+	if err := c.WaitForCluster(end); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Cluster) LoadImage() error {
+	ctx := context.Background()
+	file, err := os.Open(c.nodeImageFile)
+	if err != nil {
+		return err
+	}
+	reader := bufio.NewReader(file)
+	resp, err := c.docker.ImageLoad(ctx, reader, true)
+	defer resp.Body.Close()
+	return err
+}
+
+// Create makes all docker objects in the dind cluster.
+func (c *Cluster) CreateDockerObjects() error {
+	// First, create the docker network.
+	if err := c.createNetwork(); err != nil {
+		return err
+	}
+
+	// Create the master node.
+	if err := c.createMaster(); err != nil {
+		return err
+	}
+
+	// Create the worker nodes
+	for i := 0; i < c.numNodes-1; i++ {
+		// There are only 255 IPs in the block, so we can't exceed that. This needs to be pulled out for ipv6.
+		suffix := i + 3
+		if suffix > 255 {
+			return fmt.Errorf("Too many nodes, cannot allocate IPs")
+		}
+		ip := fmt.Sprintf("172.18.0.%d", suffix)
```

Comment:
what's `172.18.0.`?

Reply:
Atm, 172.18.0.0/24 is a subnet allocated to the docker network we're currently using.

I've some work to allow multiple configurations of the docker network, but this PR is already kind of big.